framework/include/utils/InputParameters.h

```diff
@@ -987,6 +987,31 @@ class InputParameters : public Parameters
    */
   void setParameters() {}
 
+  /**
+   * Inner method for \c addRequiredParam
+   */
+  template <typename T>
+  void addRequiredParamInner(const std::string & name, const std::string & doc_string);
```

Comment:
ugh this is quite verbose for a simple thing. 

Can you try using std::same_if and see if that works?

Reply:
Works